pkg/autodiscovery/autoconfig.go

```diff
@@ -540,3 +628,47 @@ func GetConfigErrors() map[string]string {
 func GetResolveWarnings() map[string][]string {
 	return errorStats.getResolveWarnings()
 }
+
+// processNewService takes a service, tries to match it against templates and
+// triggers scheduling events if it finds a valid config for it.
+func (ac *AutoConfig) processNewService(svc listeners.Service) {
+	// in any case, register the service and store its tag hash
+	ac.store.setServiceForEntity(svc, svc.GetEntity())
+	// todo store?
+
+	// get all the templates matching service identifiers
+	var templates []integration.Config
+	ADIdentifiers, err := svc.GetADIdentifiers()
+	if err != nil {
+		log.Errorf("Failed to get AD identifiers for service %s, it will not be monitored - %s", svc.GetEntity(), err)
+		return
+	}
+	for _, adID := range ADIdentifiers {
+		// map the AD identifier to this service for reverse lookup
+		ac.store.setADIDForServices(adID, svc.GetEntity())
+		tpls, err := ac.store.templateCache.Get(adID)
+		if err != nil {
+			log.Debugf("Unable to fetch templates from the cache: %v", err)
+		}
+		templates = append(templates, tpls...)
+	}
+	for _, template := range templates {
+		// resolve the template
+		resolvedConfig, err := ac.resolveTemplateForService(template, svc)
+		if err != nil {
+			continue
+		}
+
+		// ask the Collector to schedule the checks
+		ac.schedule([]integration.Config{resolvedConfig})
+	}
+}
+
+// processDelService takes a service, stops its associated checks, and updates the cache
+func (ac *AutoConfig) processDelService(svc listeners.Service) {
+	ac.store.removeServiceForEntity(svc.GetEntity())
```

Comment:
before, we used to lock the whole `configResolver` before removing all these different things? Is it safe to lock each removal individually now?

Reply:
Yes this should be safe thanks to the `store` which wrap data access with locks